specs/rollup-node.md

```diff
@@ -373,43 +373,38 @@ The [previous section on L2 chain derivation][l2-chain-derivation] assumes linea
-1. (Initialization) Set the initial `refL2` to the head block of the L2 execution engine.
-2. Set `parentL2` to `refL2`'s parent block and `refL1` to the L1 block that `refL2` was derived from.
-3. Fetch `currentL1`, the canonical L1 block at the same height as `refL1`.
-
-- If `currentL1 == refL1`, then `refL2` was built on a canonical L1 block:
-  - Find the next L1 block (it may not exist yet) and return `(refL2, nextRefL1)` as the starting point of the
-    re-derivation.
-    - It is necessary to ensure that no L1 re-org occurred during this lookup, i.e. that `nextRefL1.parent == refL1`.
-    - If the next L1 block does not exist yet, there is no re-org, and nothing new to derive, and we can abort the
-        process.
-- Otherwise, if `refL2` is the L2 genesis block, we have re-orged past the genesis block, which is an error that
-  requires a re-genesis of the L2 chain to fix (i.e. creating a new genesis configuration) (\*)
-- Otherwise, if either `currentL1` does not exist, or `currentL1 != refL1`, set `refL2` to `parentL2` and restart this
-  algorithm from step 2.
-  - Note: if `currentL1` does not exist, it means we are in a re-org to a shorter L1 chain.
-  - Note: as an optimization, we can cache `currentL1` and reuse it as the next value of `nextRefL1` to avoid an
-        extra lookup.
+ancestor, and can re-derive the L2 chain from that L1 block and onward.
+
+The rollup node maintains two heads of the L2 Chain: the unsafe head (often called head) and the safe head.
+Each L2 block has an L1 Origin block that it references in the L1 Info deposit transaction.
+The unsafe head is the L2 block with the highest number who's L1 Origin block is canonical in the new L1 chain.
+The safe head is the L2 block that was derived from a sequencing window that did not change with the reorg.
+
+Determining the unsafe head in the case of an L1 reorg is easy. Start with the current L2 Head and walk the L1 chain
+until the L1 Origin of an L2 block is canonical.
+
+To determine the safe head, the straightforward approach is to walk back from the unsafe head until the L1 Origin number
+is a full sequencing window behind the L1 Origin number of the unsafe head.
+
+The purpose of this is to ensure that if the sequencing window for a L2 block has changed since it was derived, that
+the L2 block is re-derived. The start of the sequencing window for a L2 block is committed to in the L1 Origin; however
+the end of the window is based on a block number (note that it cannot be done any other way). As such we must find the
+highest L1 block that is common to both L1 chains (the reorg base) and the safe head is the L2 block who's sequencing
+window ends at the reorg base. Any L2 block after that will have a different sequencing window.
+
+Walking the L2 chain until we find a canonical L1 block as the L1 origin is a straightforward way to determining the
+reorg base and it also serves as an implicit commitment to the end of the sequencing window for any L2 block.
```

Comment:
I find this paragraph from "The start of the sequencing window" and onwards to be very confusing.

> The start of the sequencing window for a L2 block is committed to in the L1 Origin

I guess you mean that the L1 origin IS the first block of the sequencing window for the epoch with the same number?

> the end of the window is based on a block number

Do you just mean that the end of the window is L1Origin + SEQUENCING_WINDOW_SIZE?

> (note that it cannot be done any other way)

If you're not going to explain why that's just unnecessary teasing :D

Reply:
> As such we must find the highest L1 block that is common to both L1 chains (the reorg base)

I see what you mean but "both L1 chains" is a bit weird. And I got confused because of the sentence structure: "both L1 chains AND ...."

> and the safe head is the L2 block who's sequencing window ends at the reorg base.

Drop the "and" and make this a new sentence:

> The safe head is the last L2 block of the epoch whose sequencing window ends at the reorg base.

Note that this is sort of already said above.

> Walking the L2 chain until we find a canonical L1 block as the L1 origin is a straightforward way to determining the
reorg base and it also serves as an implicit commitment to the end of the sequencing window for any L2 block.

This repeats what has been said a bit before, but more confusingly, so I would just drop it.

Reply:
> I guess you mean that the L1 origin IS the first block of the sequencing window for the epoch with the same number?

yes

> end of the window is L1Origin + SEQUENCING_WINDOW_SIZE

(and that it is by number, not by hash while the start is by hash)

> If you're not going to explain why that's just unnecessary teasing :D

It's impossible for the sequencer to know the end of the sequencing window by hash as it's a block in the future (and they must include a transaction that changes the hash of the block)

Reply:
Will go through and clean this up